libsolidity/analysis/ControlFlowAnalyzer.cpp

```diff
@@ -191,6 +247,100 @@ void ControlFlowAnalyzer::checkUnreachable(CFGNode const* _entry, CFGNode const*
 		// Extend the location, as long as the next location overlaps (unreachable is sorted).
 		for (; it != unreachable.end() && it->start <= location.end; ++it)
 			location.end = std::max(location.end, it->end);
-		m_errorReporter.warning(5740_error, location, "Unreachable code.");
+
+		if (m_previousUnreachable.emplace(location).second)
+			m_errorReporter.warning(5740_error, location, "Unreachable code.");
 	}
 }
+
+bool ControlFlowAnalyzer::checkForReverts(ContractDefinition const& _contract, FunctionCall const& _functionCall)
+{
+	auto const& functionType = dynamic_cast<FunctionType const&>(
+		*_functionCall.expression().annotation().type
+	);
+
+	if (!functionType.hasDeclaration())
+		return false;
+
+	auto const& unresolvedFunctionDefinition =
+		dynamic_cast<FunctionDefinition const&>(functionType.declaration());
+
+	FunctionDefinition const* functionDefinition = nullptr;
+	ContractDefinition const* contractDefinition = nullptr;
+
+	SimpleASTVisitor visitor([&](ASTNode const& _node) {
+		if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(&_node))
+		{
+			if (*memberAccess->annotation().requiredLookup == VirtualLookup::Super)
+			{
+				if (auto const typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
+					if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+					{
+						solAssert(contractType->isSuper(), "");
+						ContractDefinition const* superContract = contractType->contractDefinition().superContract(_contract);
+
+						functionDefinition =
+							&unresolvedFunctionDefinition.resolveVirtual(
+								_contract,
+								superContract
+							);
+					}
+			}
+			else
+			{
+				solAssert(*memberAccess->annotation().requiredLookup == VirtualLookup::Static, "");
+				functionDefinition = &unresolvedFunctionDefinition;
+			}
+			return false;
+		}
+		else if (auto const* identifier = dynamic_cast<Identifier const*>(&_node))
+		{
+			solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+			functionDefinition = &unresolvedFunctionDefinition.resolveVirtual(_contract);
+
+			return false;
+		}
+		return true;
+	}, [](ASTNode const&){});
+
+	_functionCall.expression().accept(visitor);
+
+	solAssert(functionDefinition != nullptr, "");
+
+	if (!functionDefinition->isFree())
+		contractDefinition = functionDefinition->annotation().contract;
+
+	auto functionKey = pair{contractDefinition, functionDefinition};
+
+	if (auto reverts = m_functionReverts.find(functionKey); reverts != m_functionReverts.end())
+		return reverts->second != RevertState::NoRevert;
```

Comment:
I find code like this much easier to read:
```suggestion
	auto reverts = m_functionReverts.find(functionKey)
	if (reverts != m_functionReverts.end())
		return reverts->second != RevertState::NoRevert;
```